# Here we'll do a baseline test on 6 Neural Network model variations on the cleaned data

In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2
import time

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import mean_squared_error
from pathlib import Path
import tensorflow as tf
import matplotlib.pyplot as plt

In [2]:
# Download raw data from postgres for stage 1 ETL

conn_string = 'postgres://whnpmxwsiccrtg:53c453893549d2b1e6a4ff92e626a2a08ebcaff66678e50d33e3742f66e3e4f4@ec2-52-4-171-132.compute-1.amazonaws.com/d2ajro4cjr10lb'

db = create_engine(conn_string)
conn = db.connect()

start_time = time.time()
df = pd.read_sql_query('select * from "merged_no_cal"',con=conn)
print("PostGres Download Duration: {} seconds".format(time.time() - start_time))
conn.close ()

PostGres Download Duration: 1.8421802520751953 seconds


In [3]:
print(df.shape)
df.head()

(4949, 256)


host_is_superhost  host_listings_count  host_identity_verified  \
0                  1                  1.0                       1   
1                  1                  2.0                       0   
2                  1                  1.0                       1   
3                  0                  1.0                       0   
4                  1                  2.0                       1   

   accommodates  bathrooms  bedrooms  price  security_deposit  cleaning_fee  \
0             2        1.0         2   40.0               0.0           0.0   
1             2        1.0         1   65.0               0.0          25.0   
2             2        1.0         1   93.0               0.0           0.0   
3             2        1.0         1  105.0             200.0           0.0   
4             2        1.0         1   55.0               0.0           0.0   

   guests_included  ...  bed_type_Couch  bed_type_Futon  \
0                1  ...             0.0             0.0   
1                1  ...             0.0             0.0   
2                1  ...             0.0             0.0   
3                1  ...             0.0             0.0   
4                1  ...             0.0             0.0   

   bed_type_Pull-out Sofa  bed_type_Real Bed  cancellation_policy_flexible  \
0                     0.0                1.0                           0.0   
1                     0.0                1.0                           0.0   
2                     0.0                1.0                           0.0   
3                     0.0                1.0                           0.0   
4                     0.0                1.0                           0.0   

   cancellation_policy_moderate  cancellation_policy_strict  \
0                           0.0                         0.0   
1                           0.0                         0.0   
2                           1.0                         0.0   
3                           0.0                         0.0   
4                           1.0                         0.0   

   cancellation_policy_strict_14_with_grace_period  \
0                                              1.0   
1                                              1.0   
2                                              0.0   
3                                              1.0   
4                                              0.0   

   cancellation_policy_super_strict_30  cancellation_policy_super_strict_60  
0                                  0.0                                  0.0  
1                                  0.0                                  0.0  
2                                  0.0                                  0.0  
3                                  0.0                                  0.0  
4                                  0.0                                  0.0  

[5 rows x 256 columns]

In [4]:
# Split our preprocessed data into our features and target arrays
y = df["price"].values
X = df.drop("price",1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [5]:
df.columns

Index(['host_is_superhost', 'host_listings_count', 'host_identity_verified',
       'accommodates', 'bathrooms', 'bedrooms', 'price', 'security_deposit',
       'cleaning_fee', 'guests_included',
       ...
       'bed_type_Couch', 'bed_type_Futon', 'bed_type_Pull-out Sofa',
       'bed_type_Real Bed', 'cancellation_policy_flexible',
       'cancellation_policy_moderate', 'cancellation_policy_strict',
       'cancellation_policy_strict_14_with_grace_period',
       'cancellation_policy_super_strict_30',
       'cancellation_policy_super_strict_60'],
      dtype='object', length=256)

In [6]:
# scaler = RobustScaler()

from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()

# Fit the scaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


# Test 1: Use one input layer and one hidden layer

In [7]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 80
hidden_nodes_layer2 = 30

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 80)                20480     
_________________________________________________________________
dense_1 (Dense)              (None, 30)                2430      
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 31        
Total params: 22,941
Trainable params: 22,941
Non-trainable params: 0
_________________________________________________________________


In [8]:
# Compile the model
nn.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

In [9]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
116/116 [==============================] - 1s 889us/step - loss: 183.8592 - mean_absolute_error: 183.8592
Epoch 2/50
116/116 [==============================] - 0s 1ms/step - loss: 90.0112 - mean_absolute_error: 90.0112
Epoch 3/50
116/116 [==============================] - 0s 856us/step - loss: 80.3731 - mean_absolute_error: 80.3731
Epoch 4/50
116/116 [==============================] - 0s 857us/step - loss: 72.7551 - mean_absolute_error: 72.7551
Epoch 5/50
116/116 [==============================] - 0s 1ms/step - loss: 72.5555 - mean_absolute_error: 72.5555
Epoch 6/50
116/116 [==============================] - 0s 911us/step - loss: 69.1262 - mean_absolute_error: 69.1262
Epoch 7/50
116/116 [==============================] - 0s 1ms/step - loss: 68.0666 - mean_absolute_error: 68.0666
Epoch 8/50
116/116 [==============================] - 0s 795us/step - loss: 68.6957 - mean_absolute_error: 68.6957
Epoch 9/50
116/116 [==============================] - 0s 843us/step - loss: 67.6753 

In [10]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}")
print(f"Mean Absolute Error: {model_accuracy}")

39/39 - 0s - loss: 64.0974 - mean_absolute_error: 64.0974
Loss: 64.09737396240234
Mean Absolute Error: 64.09737396240234


# Test 2.  Add more neurons to each layer

In [11]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50

nn2 = tf.keras.models.Sequential()

# First hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn2.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 100)               25600     
_________________________________________________________________
dense_4 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_5 (Dense)              (None, 1)                 51        
Total params: 30,701
Trainable params: 30,701
Non-trainable params: 0
_________________________________________________________________


In [12]:
# Compile the model
nn2.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model2 = nn2.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
116/116 [==============================] - 0s 1ms/step - loss: 179.9674 - mean_absolute_error: 179.9674
Epoch 2/50
116/116 [==============================] - 0s 911us/step - loss: 85.5946 - mean_absolute_error: 85.5946
Epoch 3/50
116/116 [==============================] - 0s 985us/step - loss: 75.7549 - mean_absolute_error: 75.7549
Epoch 4/50
116/116 [==============================] - 0s 891us/step - loss: 71.3882 - mean_absolute_error: 71.3882
Epoch 5/50
116/116 [==============================] - 0s 862us/step - loss: 70.4062 - mean_absolute_error: 70.4062
Epoch 6/50
116/116 [==============================] - 0s 861us/step - loss: 66.6180 - mean_absolute_error: 66.6180
Epoch 7/50
116/116 [==============================] - 0s 940us/step - loss: 63.7886 - mean_absolute_error: 63.7886
Epoch 8/50
116/116 [==============================] - 0s 870us/step - loss: 63.7697 - mean_absolute_error: 63.7697
Epoch 9/50
116/116 [==============================] - 0s 893us/step - loss: 61.5

In [13]:
# Evaluate the model using the test data
model_loss2, model_accuracy2 = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss2}")
print(f"Mean Absolute Error: {model_accuracy2}")

39/39 - 0s - loss: 62.5939 - mean_absolute_error: 62.5939
Loss: 62.5938835144043
Mean Absolute Error: 62.5938835144043


# Test 3:  Add an additional layer with 20 neurons

In [14]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20

nn3 = tf.keras.models.Sequential()

# First hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn3.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 100)               25600     
_________________________________________________________________
dense_7 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_8 (Dense)              (None, 20)                1020      
_________________________________________________________________
dense_9 (Dense)              (None, 1)                 21        
Total params: 31,691
Trainable params: 31,691
Non-trainable params: 0
_________________________________________________________________


In [15]:
# Compile the model
nn3.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model3 = nn3.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
116/116 [==============================] - 0s 1ms/step - loss: 176.6241 - mean_absolute_error: 176.6241
Epoch 2/50
116/116 [==============================] - 0s 1ms/step - loss: 81.9639 - mean_absolute_error: 81.9639
Epoch 3/50
116/116 [==============================] - 0s 1ms/step - loss: 71.6587 - mean_absolute_error: 71.6587
Epoch 4/50
116/116 [==============================] - 0s 972us/step - loss: 66.6524 - mean_absolute_error: 66.6524
Epoch 5/50
116/116 [==============================] - 0s 1ms/step - loss: 66.1766 - mean_absolute_error: 66.1766
Epoch 6/50
116/116 [==============================] - 0s 1000us/step - loss: 64.1335 - mean_absolute_error: 64.1335
Epoch 7/50
116/116 [==============================] - 0s 1ms/step - loss: 66.8806 - mean_absolute_error: 66.8806
Epoch 8/50
116/116 [==============================] - 0s 928us/step - loss: 67.2354 - mean_absolute_error: 67.2354
Epoch 9/50
116/116 [==============================] - 0s 1ms/step - loss: 60.9564 - mea

In [16]:
# Evaluate the model using the test data
model_loss3, model_accuracy3 = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss3}")
print(f"Mean Absolute Error: {model_accuracy3}")

39/39 - 0s - loss: 62.3583 - mean_absolute_error: 62.3583
Loss: 62.358253479003906
Mean Absolute Error: 62.358253479003906


# Test 4:  Add 50 neurons to the first layer and a 4th layer with 10 neurons

In [17]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 150
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10

nn4 = tf.keras.models.Sequential()

# First hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Output layer
nn4.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn4.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_10 (Dense)             (None, 150)               38400     
_________________________________________________________________
dense_11 (Dense)             (None, 50)                7550      
_________________________________________________________________
dense_12 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_13 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_14 (Dense)             (None, 1)                 11        
Total params: 47,191
Trainable params: 47,191
Non-trainable params: 0
_________________________________________________________________


In [18]:
# Compile the model
nn4.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model4 = nn4.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
116/116 [==============================] - 1s 1ms/step - loss: 180.2736 - mean_absolute_error: 180.2736
Epoch 2/50
116/116 [==============================] - 0s 1ms/step - loss: 82.1567 - mean_absolute_error: 82.1567
Epoch 3/50
116/116 [==============================] - 0s 1ms/step - loss: 72.6155 - mean_absolute_error: 72.6155
Epoch 4/50
116/116 [==============================] - 0s 1ms/step - loss: 71.0339 - mean_absolute_error: 71.0339
Epoch 5/50
116/116 [==============================] - 0s 1ms/step - loss: 68.7225 - mean_absolute_error: 68.7225
Epoch 6/50
116/116 [==============================] - 0s 1ms/step - loss: 67.6376 - mean_absolute_error: 67.6376
Epoch 7/50
116/116 [==============================] - 0s 1ms/step - loss: 63.3419 - mean_absolute_error: 63.3419
Epoch 8/50
116/116 [==============================] - 0s 1ms/step - loss: 61.9617 - mean_absolute_error: 61.9617
Epoch 9/50
116/116 [==============================] - 0s 1ms/step - loss: 61.2553 - mean_absol

In [19]:
# Evaluate the model using the test data
model_loss4, model_accuracy4 = nn4.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss4}")
print(f"Mean Absolute Error: {model_accuracy4}")

39/39 - 0s - loss: 63.1860 - mean_absolute_error: 63.1860
Loss: 63.18601608276367
Mean Absolute Error: 63.18601608276367


# Test 5:  Add 50 more neurons to the first layer and add a 5th layer with just 5 neurons

In [20]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 200
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10
hidden_nodes_layer5 = 5

nn5 = tf.keras.models.Sequential()

# First hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Fifth hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer5,kernel_initializer='normal', activation="relu"))


# Output layer
nn5.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn5.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 200)               51200     
_________________________________________________________________
dense_16 (Dense)             (None, 50)                10050     
_________________________________________________________________
dense_17 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_18 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_19 (Dense)             (None, 5)                 55        
_________________________________________________________________
dense_20 (Dense)             (None, 1)                 6         
Total params: 62,541
Trainable params: 62,541
Non-trainable params: 0
__________________________________________________

In [21]:
# Compile the model
nn5.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model5 = nn5.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
116/116 [==============================] - 1s 1ms/step - loss: 193.7341 - mean_absolute_error: 193.7341
Epoch 2/50
116/116 [==============================] - 0s 1ms/step - loss: 85.6380 - mean_absolute_error: 85.6380
Epoch 3/50
116/116 [==============================] - 0s 1ms/step - loss: 75.8783 - mean_absolute_error: 75.8783
Epoch 4/50
116/116 [==============================] - 0s 1ms/step - loss: 70.6532 - mean_absolute_error: 70.6532
Epoch 5/50
116/116 [==============================] - 0s 1ms/step - loss: 66.0273 - mean_absolute_error: 66.0273
Epoch 6/50
116/116 [==============================] - 0s 1ms/step - loss: 66.9521 - mean_absolute_error: 66.9521
Epoch 7/50
116/116 [==============================] - 0s 1ms/step - loss: 65.7451 - mean_absolute_error: 65.7451
Epoch 8/50
116/116 [==============================] - 0s 1ms/step - loss: 62.2516 - mean_absolute_error: 62.2516
Epoch 9/50
116/116 [==============================] - 0s 1ms/step - loss: 64.1361 - mean_absol

In [22]:
# Evaluate the model using the test data
model_loss5, model_accuracy5 = nn5.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss5}")
print(f"Mean Absolute Error: {model_accuracy5}")

39/39 - 0s - loss: 63.3892 - mean_absolute_error: 63.3892
Loss: 63.38916778564453
Mean Absolute Error: 63.38916778564453


# Test 6:  Add 150 neurons to first layer and change second activation function to tanh

In [23]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 350
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10
hidden_nodes_layer5 = 5

nn6 = tf.keras.models.Sequential()

# First hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="tanh"))

# Third hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Fifth hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer5,kernel_initializer='normal', activation="relu"))

# Output layer
nn6.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn6.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_21 (Dense)             (None, 350)               89600     
_________________________________________________________________
dense_22 (Dense)             (None, 50)                17550     
_________________________________________________________________
dense_23 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_24 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_25 (Dense)             (None, 5)                 55        
_________________________________________________________________
dense_26 (Dense)             (None, 1)                 6         
Total params: 108,441
Trainable params: 108,441
Non-trainable params: 0
________________________________________________

In [24]:
# Compile the model
nn6.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model6 = nn6.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
116/116 [==============================] - 1s 1ms/step - loss: 200.1604 - mean_absolute_error: 200.1604
Epoch 2/50
116/116 [==============================] - 0s 1ms/step - loss: 141.5072 - mean_absolute_error: 141.5072
Epoch 3/50
116/116 [==============================] - 0s 1ms/step - loss: 99.2916 - mean_absolute_error: 99.2916
Epoch 4/50
116/116 [==============================] - 0s 2ms/step - loss: 76.8852 - mean_absolute_error: 76.8852
Epoch 5/50
116/116 [==============================] - 0s 1ms/step - loss: 64.6838 - mean_absolute_error: 64.6838
Epoch 6/50
116/116 [==============================] - 0s 1ms/step - loss: 61.4869 - mean_absolute_error: 61.4869
Epoch 7/50
116/116 [==============================] - 0s 1ms/step - loss: 54.6593 - mean_absolute_error: 54.6593
Epoch 8/50
116/116 [==============================] - 0s 1ms/step - loss: 52.4966 - mean_absolute_error: 52.4966
Epoch 9/50
116/116 [==============================] - 0s 1ms/step - loss: 50.1934 - mean_abs

In [25]:
# Evaluate the model using the test data
model_loss6, model_accuracy6 = nn6.evaluate(X_train_scaled,y_train,verbose=2)
print(f"Loss: {model_loss6}")
print(f"Mean Absolute Error: {model_accuracy6}")

116/116 - 0s - loss: 12.4532 - mean_absolute_error: 12.4532
Loss: 12.45322036743164
Mean Absolute Error: 12.45322036743164


In [26]:
# Evaluate the model using the test data
model_loss6, model_accuracy6 = nn6.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss6}")
print(f"Mean Absolute Error: {model_accuracy6}")

39/39 - 0s - loss: 61.0384 - mean_absolute_error: 61.0384
Loss: 61.03842544555664
Mean Absolute Error: 61.03842544555664


# After lots of tests, our best Mean Squared Error Result is off, on average, by about $61.04 (according to the MAE)